File: internal/wire/encoder.go
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package wire

import (
	"encoding/binary"
	"fmt"
	"io"
)

type Encoder struct {
	w          io.Writer
	maxPayload int
}

type EncoderOption func(*Encoder)

func WithChunkSize(n int) EncoderOption {
	return func(e *Encoder) { e.maxPayload = n }
}

func NewEncoder(w io.Writer, opts ...EncoderOption) *Encoder {
	e := &Encoder{w: w, maxPayload: MaxPayloadBytes}
	for _, opt := range opts {
		opt(e)
	}
	return e
}

func (e *Encoder) Encode(m *Message) error {
	if !m.Type.Valid() {
		return fmt.Errorf("%w: %d", ErrUnknownType, m.Type)
	}

	p := m.Payload
	total := 1
	if len(p) > e.maxPayload {
		if m.Type.isControl() {
			return fmt.Errorf("%w: payload %d > %d", ErrControlMultiChunk, len(p), e.maxPayload)
		}
		total = (len(p) + e.maxPayload - 1) / e.maxPayload
		if total > MaxChunks {
			return fmt.Errorf("%w: payload %d would need %d chunks > MaxChunks %d",
				ErrMessageTooLarge, len(p), total, MaxChunks)
		}
	}

	for i := 0; i < total; i++ {
		start := i * e.maxPayload
		end := min(start+e.maxPayload, len(p))
		if err := e.writeChunk(m, uint32(i), uint32(total), p[start:end]); err != nil {
			return err
		}
	}
	return nil
}

func (e *Encoder) writeChunk(m *Message, idx, total uint32, payload []byte) error {
	var hdr [4 + headerSize]byte
	binary.BigEndian.PutUint32(hdr[0:4], uint32(headerSize+len(payload)))
	hdr[4] = byte(m.Type)
	binary.BigEndian.PutUint64(hdr[5:13], m.ReqID)
	binary.BigEndian.PutUint32(hdr[13:17], m.RegionID)
	hdr[17] = m.HookID
	hdr[18] = 0 // chunk_flags reserved for future use (compression, etc.)
	binary.BigEndian.PutUint32(hdr[19:23], idx)
	binary.BigEndian.PutUint32(hdr[23:27], total)

	if _, err := e.w.Write(hdr[:]); err != nil {
		return err
	}
	if len(payload) > 0 {
		if _, err := e.w.Write(payload); err != nil {
			return err
		}
	}
	return nil
}

File: java/com/virogg/hbasecop/bridge/wire/Encoder.java
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package com.virogg.hbasecop.bridge.wire;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Serializes a {@link Message} into a single {@link ByteBuffer} containing one or more chunk frames
 * back-to-back. Mirrors {@code internal/wire/encoder.go}.
 *
 * <p>The Encoder is stateless; a single instance may be reused across threads only if callers
 * supply distinct destination buffers per call (the encoder itself does not retain state).
 */
public final class Encoder {

  private final int maxPayload;

  /** Default Encoder using {@link WireFormat#MAX_PAYLOAD_BYTES} per chunk. */
  public Encoder() {
    this(WireFormat.MAX_PAYLOAD_BYTES);
  }

  /**
   * Encoder that splits payloads into chunks of at most {@code chunkSize} bytes. Used by tests to
   * exercise the chunking path without 64 KiB allocations; production code should use the
   * no-argument constructor.
   */
  public Encoder(int chunkSize) {
    if (chunkSize <= 0) {
      throw new IllegalArgumentException("chunkSize must be > 0: " + chunkSize);
    }
    this.maxPayload = chunkSize;
  }

  /**
   * Encode {@code m} into a fresh ByteBuffer (already flipped, ready to read). For multi-chunk
   * messages the buffer contains every chunk in chunk_idx order.
   */
  public ByteBuffer encode(Message m) throws WireException {
    if (!m.type().valid()) {
      throw new UnknownTypeException("type=" + m.type());
    }

    byte[] payload = m.payload();
    int total = 1;
    if (payload.length > maxPayload) {
      if (m.type().isControl()) {
        throw new ControlMultiChunkException(
            "control payload " + payload.length + " > maxPayload " + maxPayload);
      }
      total = (payload.length + maxPayload - 1) / maxPayload;
      // Refuse to emit a frame stream the matching Decoder would reject
      // (chunk_total > MAX_CHUNKS). Fail at the producer with the payload
      // size in hand rather than producing self-undecodable output; this
      // also keeps the totalBytes computation below from overflowing int.
      if (total > WireFormat.MAX_CHUNKS) {
        throw new MessageTooLargeException(
            "payload "
                + payload.length
                + " would need "
                + total
                + " chunks > MAX_CHUNKS "
                + WireFormat.MAX_CHUNKS);
      }
    }

    int totalBytes = total * (4 + WireFormat.HEADER_SIZE) + payload.length;
    ByteBuffer buf = ByteBuffer.allocate(totalBytes).order(ByteOrder.BIG_ENDIAN);

    for (int i = 0; i < total; i++) {
      int start = i * maxPayload;
      int end = Math.min(start + maxPayload, payload.length);
      writeChunk(buf, m, i, total, payload, start, end);
    }

    buf.flip();
    return buf;
  }

  private static void writeChunk(
      ByteBuffer buf, Message m, int idx, int total, byte[] payload, int start, int end) {
    int payloadLen = end - start;
    buf.putInt(WireFormat.HEADER_SIZE + payloadLen);
    buf.put(m.type().value());
    buf.putLong(m.reqId());
    buf.putInt(m.regionId());
    buf.put(m.hookId());
    buf.put((byte) 0); // chunk_flags reserved
    buf.putInt(idx);
    buf.putInt(total);
    if (payloadLen > 0) {
      buf.put(payload, start, payloadLen);
    }
  }
}
